api/censusdb/censusdb.go

```diff
@@ -0,0 +1,199 @@
+// Authentication is checked if authToken is not nil.
+func (c *CensusDB) Load(censusID []byte, authToken *uuid.UUID) (*CensusRef, error) {
+	// if the tree is in memory, just return it
+	val, ok := c.censusMap.Load(string(censusID))
+	if ok {
+		ref, refOk := val.(CensusRef)
+		if !refOk {
+			panic("stored value is not of censusRef type")
+		}
+		if authToken != nil {
+			if ref.AuthToken == nil {
+				return nil, fmt.Errorf("census is locked")
+			}
+			if !bytes.Equal(authToken[:], ref.AuthToken[:]) {
+				return nil, fmt.Errorf("wrong authentication token")
+			}
+		}
+		return &ref, nil
+	}
+
+	// if not in memory, load census from DB
+	ref, err := c.getCensusRefFromDB(censusID)
+	if err != nil {
+		return nil, err
+	}
+	// check authentication
+	if authToken != nil {
+		// if no token stored in the reference but the called provided a token, we don't allow
+		if ref.AuthToken == nil {
+			return nil, fmt.Errorf("census is locked")
+		}
+		if !bytes.Equal(authToken[:], ref.AuthToken[:]) {
+			return nil, fmt.Errorf("wrong authentication token")
+		}
+	}
+	ref.tree, err = censustree.New(censustree.Options{Name: censusName(censusID), ParentDB: c.db,
+		MaxLevels: 256, CensusType: models.Census_Type(ref.CensusType)})
+	if err != nil {
+		return nil, err
+	}
+	if ref.IsPublic {
+		ref.tree.Publish()
+	}
+	c.censusMap.Store(string(censusID), *ref)
+	log.Debugf("loaded tree %x", censusID)
+	return ref, nil
+}
+
+// addCensusRefToDB adds a censusRef to the database.
+func (c *CensusDB) addCensusRefToDB(censusID []byte, authToken *uuid.UUID,
+	t models.Census_Type, indexed, public bool) (*CensusRef, error) {
+	c.censusDBlock.Lock()
+	defer c.censusDBlock.Unlock()
+	wtx := c.db.WriteTx()
+	defer wtx.Discard()
+	refData := bytes.Buffer{}
+	enc := gob.NewEncoder(&refData)
+	ref := &CensusRef{
+		AuthToken:  authToken,
+		CensusType: int32(t),
+		Indexed:    indexed,
+		IsPublic:   public,
+	}
+	if err := enc.Encode(ref); err != nil {
+		return nil, err
+	}
+	if err := wtx.Set(append([]byte(censusDBreferencePrefix), censusID...),
+		refData.Bytes()); err != nil {
+		return nil, err
+	}
+	return ref, wtx.Commit()
+}
+
+// Del removes a census from the database and memory.
+func (c *CensusDB) Del(censusID []byte) error {
+	c.censusDBlock.Lock()
+	defer c.censusDBlock.Unlock()
+	wtx := c.db.WriteTx()
+	defer wtx.Discard()
+	if err := wtx.Delete(append([]byte(censusDBreferencePrefix), censusID...)); err != nil {
+		return err
+	}
+	// the removal of the tree from the disk is done in a separate goroutine.
+	// This is because the tree is locked and we don't want to block the operations,
+	// and depending on the size of the tree, it can take a while to delete it.
+	go func() {
+		c.censusMap.Delete(string(censusID))
+		_, err := censustree.DeleteCensusTreeFromDatabase(c.db, censusName(censusID))
+		if err != nil {
+			log.Warnf("error deleting census tree %x: %s", censusID, err)
+		}
+	}()
```

Comment:
If we keep pebble alone, without mutexes, then I would delete this goroutine. you would not be preventing other operations by keeping the transaction open for a little longer, to delete the entire census tree. you'd also save work - you'd use one goroutine and one transaction, whereas now you use two goroutines and two transactions.

Reply:
The Delete operation might take many seconds (depending on the size of the census) since the only way to achieve it is by iterating over all leaves. This goroutine ensures the HTTP request is replied before any timeout is triggered. And once the reference is removed (that's executed before launching the routine) at practical effects the census is not available anymore for the API.